cmd/dockerd/daemon_unix.go

```diff
@@ -130,33 +122,50 @@ func newCgroupParent(config *config.Config) string {
 	return cgroupParent
 }
 
-func (cli *DaemonCli) initContainerD(ctx context.Context) (func(time.Duration) error, error) {
-	var waitForShutdown func(time.Duration) error
-	if cli.Config.ContainerdAddr == "" {
-		systemContainerdAddr, ok, err := systemContainerdRunning(honorXDG)
+func (cli *DaemonCli) initContainerd(ctx context.Context) (func(), error) {
+	if cli.Config.ContainerdAddr != "" {
+		// use system containerd at the given address
+		return nil, nil
+	}
+
+	systemContainerdAddr, ok, err := systemContainerdRunning(honorXDG)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not determine whether the system containerd is running")
+	}
+	if ok {
+		// detected a system containerd at the given address
+		cli.Config.ContainerdAddr = systemContainerdAddr
+		return nil, nil
+	}
+
+	logrus.Debug("containerd not running, starting daemon managed containerd")
+	opts := []supervisor.DaemonOpt{
+		supervisor.WithOOMScore(cli.Config.OOMScoreAdjust),
+	}
+
+	if cli.Config.Debug {
+		opts = append(opts, supervisor.WithLogLevel("debug"))
+	} else if cli.Config.LogLevel != "" {
+		opts = append(opts, supervisor.WithLogLevel(cli.Config.LogLevel))
+	}
+
+	if !cli.Config.CriContainerd {
+		opts = append(opts, supervisor.WithPlugin("cri", nil))
+	}
+
+	r, err := supervisor.Start(ctx, filepath.Join(cli.Config.Root, "containerd"), filepath.Join(cli.Config.ExecRoot, "containerd"), opts...)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to start containerd")
+	}
+	cli.Config.ContainerdAddr = r.Address()
+	logrus.Info("started daemon managed containerd using " + cli.Config.ContainerdAddr)
```

Comment:
Nit: structured logging?
```suggestion
	logrus.WithField("address", cli.config.ContainerdAddr).Info("started daemon managed containerd")
```

Reply:
done